### PR TITLE
Do not remove temporarily failed containers

### PR DESCRIPTION
**Pull request by astarnell**
- Add Sentry
- Treat agents as state full (do not reset them every time).
- Only add new containers to agents and not all every time. With that a container is only queried once from the Kubernetes API and the state is stored in the agent.
- Files in the `containers` directory is the only source of truth to decide if a container is stall

### CODE REVIEW

File: kube_log_watcher/agents/appdynamics.py [hunk 1]
```diff
@@ -61,6 +62,7 @@ def remove_log_target(self, container_id):
             logger.debug('AppDynamics watcher agent Removed container({}) job file'.format(container_id))
         except Exception:
             logger.exception('AppDynamics watcher agent Failed to remove job file: {}'.format(job_file))
+            sentry_sdk.capture_exception()
```
Comment (anton-ryzhov) [inline kube_log_watcher/agents/appdynamics.py]: `logging.exception` does the sentry capturing by default

File: kube_log_watcher/agents/scalyr.py [hunk 2]
```diff
@@ -48,6 +49,7 @@ def container_annotation(annotations, container_name, pod_name, annotation_key,
                         return candidate.get(result_key, default)
         except Exception:
             logger.error('Scalyr watcher agent failed to load annotation {}'.format(annotation_key))
+            sentry_sdk.capture_exception()
```
Comment (anton-ryzhov) [inline kube_log_watcher/agents/scalyr.py]: `logger.error` does the sentry capturing by default

File: kube_log_watcher/agents/scalyr.py [hunk 3]
```diff
@@ -162,6 +164,7 @@ def parse_scalyr_sampling_rules(self, scalyr_sampling_rules):
                 json.loads(scalyr_sampling_rule['value'])
             except (TypeError, KeyError, ValueError) as error:
                 logger.warning('Cannot parse rule `{}`: {}'.format(scalyr_sampling_rule, repr(error)))
+                sentry_sdk.capture_exception()
```
Comment (anton-ryzhov) [inline kube_log_watcher/agents/scalyr.py]: If it is a warning, maybe we don't want to capture this

File: .python-version [hunk 4]
```diff
@@ -0,0 +1 @@
+3.5.2
```
Comment (anton-ryzhov) [inline .python-version]: Is it used anyhow?

File: kube_log_watcher/main.py [hunk 5]
```diff
@@ -281,7 +289,12 @@ def watch(containers_path, agents_list, cluster_id, interval=60, kube_url=None,
     watcher_config = load_watcher_config(watcher_config_file)
 
     configuration = dict(watcher_config, cluster_id=cluster_id)
-    agents = load_agents(agents_list, configuration)
+
+    try:
+        agents = load_agents(agents_list, configuration)
+    except Exception:
+        sentry_sdk.capture_exception()
```
Comment (anton-ryzhov) [inline kube_log_watcher/main.py]: Unhandled exception will be captured by sentry

File: kube_log_watcher/main.py [hunk 6]
```diff
@@ -295,25 +308,38 @@ def watch(containers_path, agents_list, cluster_id, interval=60, kube_url=None,
...
+            sentry_sdk.capture_exception()
             time.sleep(interval / 2)
 
 
 def main():
+    sentry_dsn = os.environ.get("SENTRY_DSN")
+    if sentry_dsn:
+        sentry_sdk.init(
+            dsn=os.environ["SENTRY_DSN"],
+            release=os.environ["VERSION"],
```
Comment (anton-ryzhov) [inline kube_log_watcher/main.py]: Potential `KeyError`, please provide some default value

File: kube_log_watcher/main.py [hunk 7]
```diff
@@ -281,7 +285,12 @@ def watch(containers_path, agents_list, cluster_id, interval=60, kube_url=None,
     watcher_config = load_watcher_config(watcher_config_file)
 
     configuration = dict(watcher_config, cluster_id=cluster_id)
-    agents = load_agents(agents_list, configuration)
+
+    try:
+        agents = load_agents(agents_list, configuration)
+    except Exception:
+        logger.exception('Failed to load agents')
```
Comment (anton-ryzhov) [inline kube_log_watcher/main.py]: Unhandled exception will be captured by sentry -> you maybe don't need this try-except at all

File: .python-version (hunk 4 above)
Comment (astarnell) [reply .python-version]: It is used by pyenv to automatically use this version in this directory

File: kube_log_watcher/agents/base.py [hunk 8]
```diff
@@ -20,11 +20,10 @@ def first_run(self):
         return True
 
     def __enter__(self):
-        self.reset()
```
Comment (anton-ryzhov) [inline kube_log_watcher/agents/base.py]: This change doesn't look safe.
How reset is executed now?

File: kube_log_watcher/main.py [hunk 9]
```diff
@@ -254,8 +258,8 @@ def get_new_containers_log_targets(
     return containers_log_targets
 
 
-def get_stale_containers(watched_containers: set, existing_container_ids: list) -> int:
-    return set(watched_containers) - set(existing_container_ids)
+def get_stale_containers(watched_containers: set, existing_container_ids: set) -> set:
```
Comment (anton-ryzhov) [inline kube_log_watcher/main.py]: I would suggest to inline this function

File: kube_log_watcher/agents/base.py (hunk 8 above)
Comment (astarnell) [reply kube_log_watcher/agents/base.py]: Probably never. That is the idea of the PR. To use the agent to save its state between two executions and only add new or remove stale containers.

Comment (anton-ryzhov) [reply kube_log_watcher/agents/base.py]: Okay, so you're relying on `remove_log_target` Then it should remove items from `.logs` for scalyr and appdynamics.

File: kube_log_watcher/main.py [hunk 10]
```diff
@@ -165,7 +169,7 @@ def sync_containers_log_agents(
             logger.exception('Failed to sync log config with agent {}'.format(agent.name))
 
     # 4. return new containers, stale containers
-    return existing_container_ids, stale_container_ids
+    return new_container_ids, set(stale_container_ids)
```
Comment (anton-ryzhov) [inline kube_log_watcher/main.py]: Isn't `stale_container_ids` a set?

File: kube_log_watcher/main.py (hunk 9 above)
Comment (astarnell) [reply kube_log_watcher/main.py]: done

File: kube_log_watcher/main.py (hunk 10 above)
Comment (astarnell) [reply kube_log_watcher/main.py]: ähm yes. Forgot to remove it after debugging

File: kube_log_watcher/agents/base.py (hunk 8 above)
Comment (astarnell) [reply kube_log_watcher/agents/base.py]: Right, remove_log_target should remove items from .logs. I missed that this was not the case. Now it removes the items.

File: .gitignore [hunk 11]
```diff
@@ -35,6 +35,8 @@ coverage.xml
 *.iml
 .cache/
 *.swp
+venv/
```
[the remark this answers is not in the document]
Comment (astarnell) [reply .gitignore]: Of cause its host-specific. But venv or .venv seem to be often used names. Even [python documentation](https://docs.python.org/3/library/venv.html#creating-virtual-environments) mentions this `(a common name for the target directory is .venv)`

**Comment by anton-ryzhov** [comment]
:+1:

**Comment by astarnell** [comment]
:+1:

**Pull request merged**